networking_cisco/plugins/cisco/l3/drivers/asr1k/aci_asr1k_routertype_driver.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2015 Cisco Systems, Inc.  All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from networking_cisco.plugins.cisco.extensions import routerhostingdevice
+from networking_cisco.plugins.cisco.l3.drivers.asr1k import (
+    asr1k_routertype_driver as asr1k)
+from neutron.common import constants as l3_constants
+from neutron import context as nctx
+from neutron import manager
+
+from oslo_log import log as logging
+from oslo_utils import importutils
+
+LOG = logging.getLogger(__name__)
+
+
+APIC_ML2_L3DRIVER_KLASS = (
+    'apic_ml2.neutron.services.l3_router.apic_driver.ApicL3Driver')
+GBP_L3DRIVER_KLASS = (
+    'gbpservice.neutron.services.l3_router.apic_driver.ApicGBPL3Driver')
+HOSTING_DEVICE_ATTR = routerhostingdevice.HOSTING_DEVICE_ATTR
+
+
+class AciASR1kL3RouterDriver(asr1k.ASR1kL3RouterDriver):
+
+    def __init__(self):
+        super(AciASR1kL3RouterDriver, self).__init__()
+        self._apic_driver = None
+
+    @property
+    def apic_driver(self):
+        """Get APIC driver
+
+        There are different drivers for the GBP workflow
+        and Neutron workflow for APIC. First see if the GBP
+        workflow is active, and if so get the APIC driver for it.
+        If the GBP service isn't installed, try to get the driver
+        from the Neutron (APIC ML2) workflow.
+        """
+        if not self._apic_driver:
+            try:
+                if manager.NeutronManager.get_service_plugins()[
+                    'GROUP_POLICY'].policy_driver_manager.policy_drivers[
+                        'apic'].obj:
+                    self._apic_driver = importutils.import_object(
+                        GBP_L3DRIVER_KLASS,
+                        nctx.get_admin_context_without_session()
+                    )
+                    self._apic_driver._plugin = self._l3_plugin
+            except Exception:
```

Comment:
if its GBP workflow but the gbp_l3driver can't be loaded, then the error msg won't be logged until later when doing if check for ml2. Plus that error msg will be ml2-specific which has nothing to do with the real problem (gbp_l3driver can't be loaded).